### PR TITLE
Add CLI option to parallelize file processing

### PR DESCRIPTION
**Pull request by or**
The files processed by the 'fix' and 'check' commands can be processed
in an arbitrary order and each file can be processed independently, so
the whole process can be parallelized.

To make sure threads aren't writing to stdout/stderr at the same time,
the result for each file is captured and then printed in order once
every file is done. The 'check' command already did this via
print-file-status, the 'fix' command now uses the same mechanism.

Addresses one part of #254.

### CODE REVIEW

**Comment by weavejester** [comment]
> To make sure threads aren't writing to stdout/stderr at the same time, the result for each file is captured and then printed in order once every file is done.

Can you explain how you're doing this? I'm not sure I'm following the logic.

**Comment by weavejester** [comment]
Ah, wait, I see. Nevermind 😃

**Comment by or** [comment]
Removed the `(flush)`. I needed that in a previous version, but no more.

Also noticed that there might be an opportunity to merge `check` and `fix`, now that they look so similar.

**Review by weavejester — CHANGES_REQUESTED** [review]
Overall good, but some changes needed.

File: cljfmt/src/cljfmt/main.clj [hunk 1]
```diff
@@ -111,34 +113,41 @@
...
+                     (map (fn [status]
+                            (print-file-status options status)
+                            (:counts status)))
+                     (reduce merge-counts))]
      (print-final-count counts)
      (exit counts))))
 
+(defn- fix-one [options file]
+  (let [original (slurp file)
+        status {:file file}]
```
[the remark this answers is not in the document]
Comment (or) [reply cljfmt/src/cljfmt/main.clj:129]: I thought `check` and `check-one` did this well, and that it makes `fix` more readable.

But I'm not married to it, if you prefer it inlined.
Would look something like this?
```clj
(defn fix
  "Applies the formatting (as per `options`) to the Clojure files
  contained in `paths`."
  ([paths]
   (fix paths {}))
  ([paths options]
   (->> paths
        (mapcat (partial find-files options))
        ((if (:parallel? options) pmap map)
         (fn [file]
           (let [original (slurp file)
                 status {:file file}]
             (try
               (let [revised (reformat-string options original)]
                 (if (not= original revised)
                   (do (spit file revised)
                       (assoc status :reformatted true))
                   status))
               (catch Exception e
                 (assoc status :exception e))))))
        (map (partial print-file-status options))
        doall)))
```

File: cljfmt/src/cljfmt/main.clj [hunk 2]
```diff
@@ -111,34 +113,41 @@
...
-                 (comp (mapcat (partial find-files options))
-                       (map (partial check-one options))
-                       (map (fn [status]
-                              (print-file-status options status)
-                              (:counts status))))
-                 (completing merge-counts)
-                 paths)]
+   (let [counts (->> paths
+                     (mapcat (partial find-files options))
+                     ((if (:parallel? options) pmap map)
```
[the remark this answers is not in the document]
Comment (or) [reply cljfmt/src/cljfmt/main.clj:118]: Ah, smart. Yes, that'll look better.

File: cljfmt/src/cljfmt/main.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (weavejester) [reply cljfmt/src/cljfmt/main.clj:129]: Ah, sorry, I wasn't clear. `fix-one` is fine and reasonable, I was thinking of `status`. So instead:

```clojure
(defn- fix-one [options file]
  (let [original (slurp file)]
    (try
      (let [revised (reformat-string options original)]
        (if (not= original revised)
          (do (spit file revised)
              {:file file, :reformatted true})
          {:file file}))
      (catch Exception ex
        {:file file, :exception ex}))))
```

Should be slightly faster as it avoids the `assoc`, and its clearer what's being returned.

[the remark this answers is not in the document]
Comment (or) [reply cljfmt/src/cljfmt/main.clj:129]: Oh, right. That was also inspired by `check-one`, so `:file` doesn't have to be added thrice.
But that's fair enough, I'll update it tomorrow morning, cheers.

File: cljfmt/src/cljfmt/main.clj [hunk 3]
```diff
@@ -111,34 +113,40 @@
...
   ([paths options]
-   (let [counts (transduce
-                 (comp (mapcat (partial find-files options))
-                       (map (partial check-one options))
-                       (map (fn [status]
-                              (print-file-status options status)
-                              (:counts status))))
-                 (completing merge-counts)
-                 paths)]
+   (let [map* (if (:parallel? options) pmap map)
```
Comment (weavejester) [inline cljfmt/src/cljfmt/main.clj:116]: Can you align the `let` terms? So:

```clojure
(let [map*   (if (:parallel? options) pmap map)
      counts (->> paths
```

Comment (or) [reply cljfmt/src/cljfmt/main.clj]: Aye, done. I'll try to keep that in mind.

**Comment by or** [comment]
Cool, thanks. This resolves #104 and #185 as well.

**Pull request merged**